extension/src/state_aggregate.rs

```diff
@@ -33,20 +33,74 @@ pub mod toolkit_experimental {
             states_len: u64, // TODO JOSH this and durations_len can be 32
             durations_len: u64,
             durations: [DurationInState; self.durations_len],
+            first_time: i64,
+            last_time: i64,
+            first_state: u32,
+            last_state: u32,  // first/last state are idx into durations, keep together for alignment
             states: [u8; self.states_len],
         }
     }
 
     impl StateAgg<'_> {
-        pub fn new(states: String, durations: Vec<DurationInState>) -> Self {
+        pub(super) fn new(
+            states: String,
+            durations: Vec<DurationInState>,
+            first: Option<Record>,
+            last: Option<Record>,
+        ) -> Self {
+            if durations.is_empty() {
+                assert!(first.is_none() && last.is_none() && states.is_empty());
+
+                return unsafe {
+                    flatten!(StateAgg {
+                        states_len: 0,
+                        states: states.into_bytes().into(),
+                        durations_len: 0,
+                        durations: (&*durations).into(),
```

Comment:
These look pretty weird, given that states and durations are both empty here. Why not `Slice::Slice(&[])`?

At that point, you're looking at a big pile of default values, which is why I suggested `impl Default`.

This doesn't need to hold up merge. But if you want, I can push a branch demonstrating what it would look like.